src/compiler/checker.ts

```diff
@@ -5667,6 +5752,191 @@ namespace ts {
+                    types.push(rspread.left);
+                }
+                types.push(rspread.right);
+                return getSpreadType(types, symbol, aliasSymbol, aliasTypeArguments);
+            }
+            if (right.flags & TypeFlags.Intersection) {
+                right = resolveObjectIntersection(right as IntersectionType);
+            }
+            if (right.flags & TypeFlags.Union) {
+                const spreads = map((right as UnionType).types,
+                                    t => getSpreadType(types.slice().concat([t]), symbol, aliasSymbol, aliasTypeArguments));
+                return getUnionType(spreads, /*subtypeReduction*/ false, aliasSymbol, aliasTypeArguments);
+            }
+            const atBeginning = types.length === 0;
+            let left = getSpreadType(types, symbol, aliasSymbol, aliasTypeArguments);
+            if (right.flags & (TypeFlags.Primitive & ~TypeFlags.StringLike) || left.flags & TypeFlags.Any) {
+                return left;
+            }
+            if (right.flags & TypeFlags.TypeParameter &&
+                left.flags & TypeFlags.Spread &&
+                (left as SpreadType).right.flags & TypeFlags.TypeParameter &&
+                right.symbol === (left as SpreadType).right.symbol) {
+                // for types like T ... T, just return ... T
+                return left;
+            }
+
+            if (right.flags & TypeFlags.ObjectType &&
+                left.flags & TypeFlags.Spread &&
+                (left as SpreadType).right.flags & TypeFlags.ObjectType) {
+                // simplify two adjacent object types: T ... { x } ... { y } becomes T ... { x, y }
+                const simplified = getSpreadType([right, (left as SpreadType).right], symbol, aliasSymbol, aliasTypeArguments);
+                return getSpreadType([(left as SpreadType).left, simplified], symbol, aliasSymbol, aliasTypeArguments);
+            }
+            if (left.flags & TypeFlags.Intersection) {
+                left = resolveObjectIntersection(left as IntersectionType);
+            }
+            if (left.flags & TypeFlags.Union) {
+                const spreads = map((left as UnionType).types,
+                                  t => getSpreadType(types.slice().concat([t, right]), symbol, aliasSymbol, aliasTypeArguments));
+                return getUnionType(spreads, /*subtypeReduction*/ false, aliasSymbol, aliasTypeArguments);
+            }
+            if (right.flags & TypeFlags.ObjectType && left.flags & TypeFlags.ObjectType) {
+                const members = createMap<Symbol>();
+                const skippedPrivateMembers = createMap<boolean>();
+                let stringIndexInfo = unionIndexInfos(getIndexInfoOfType(left, IndexKind.String), getIndexInfoOfType(right, IndexKind.String));
+                let numberIndexInfo = unionIndexInfos(getIndexInfoOfType(left, IndexKind.Number), getIndexInfoOfType(right, IndexKind.Number));

```

Comment:
only union if BOTH have indexers. If only one does, then we can't tell. Also make indexers illegal on object literal types that have spreads in them.